File: src/main/util/gui/input/SliderInput.java
package main.util.gui.input;

import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

import org.lwjgl.util.vector.Vector3f;

import com.alee.extended.panel.GroupPanel;
import com.alee.laf.label.WebLabel;
import com.alee.laf.slider.WebSlider;

public abstract class SliderInput extends GroupPanel {
	
	private WebLabel label;
	private WebSlider slider;
	private int lastValue = 0;

	public SliderInput(String labelString, int orientation, int min, int max, int value) {
		label = new WebLabel(labelString);
		slider = new WebSlider(orientation, min, max, value);
		lastValue = value;
		
		slider.addChangeListener(new ChangeListener() {
			
			@Override
			public void stateChanged(ChangeEvent e) {
				int value = slider.getValue();
				int delta = value - lastValue;
				
				WebSlider slider = (WebSlider) e.getSource();
				onValueChange(value, delta);
				
				lastValue = value;
			}
		});
		
		this.add(label);
		this.add(slider);
	}
	
	public abstract void onValueChange(int value, int delta);
}

File: engine/src/main/java/de/hanno/hpengine/util/gui/input/SliderInput.kt
package de.hanno.hpengine.util.gui.input

import com.alee.extended.panel.WebTitledPanel
import com.alee.laf.label.WebLabel
import com.alee.laf.slider.WebSlider

abstract class SliderInput @JvmOverloads constructor(labelString: String,
                           orientation: Int,
                           min: Int,
                           max: Int,
                           initialValue: Int,
                           minorTickSpacing: Int = (max - min) / 10,
                           majorTickSpacing: Int = (max-min) / 4) : WebTitledPanel() {

    private var lastValue = 0

    init {
        title = WebLabel(labelString)
        lastValue = initialValue
        val slider = WebSlider(orientation, min, max, initialValue).apply {
            paintTicks = true
            paintLabels = true

            this.minorTickSpacing = minorTickSpacing
            this.majorTickSpacing = majorTickSpacing

            addChangeListener { e ->
                val delta = value - lastValue

                onValueChange(value, delta)

                lastValue = value
            }
        }
        content = slider

    }

    abstract fun onValueChange(value: Int, delta: Int)
}
